### ISSUE TITLE
alternator,ttl: test_ttl_expiration eventually fails when ran multiple times in a loop

### ISSUE THREAD

**Issue by psarna** [scylladb/scylla#9716]
In order to test one of my patches, I left alternator TTL pytest suite running in a long loop, in order to make sure that it doesn't occasionally fail. It did fail, so I retried the experiment on master and found out that it fails in there as well.

I ran the loop below on my local Scylla instance, running on a single shard:
```bash
while pytest -vvv --runveryslow test/alternator/test_ttl.py::test_ttl_expiration; do date; done
```

After some time, ranging from 5 to 60+ seconds, I always eventually observed a failure:

```
========================================================================= test session starts ==========================================================================
platform linux -- Python 3.9.7, pytest-6.2.5, py-1.10.0, pluggy-1.0.0 -- /usr/bin/python3
cachedir: .pytest_cache
Test order randomisation NOT enabled. Enable with --random-order or --random-order-bucket=<bucket_type>
rootdir: /home/sarna/repo/scylla/test/alternator, configfile: pytest.ini
plugins: repeat-0.9.1, random-order-1.0.4
collected 1 item                                                                                                                                                       

test/alternator/test_ttl.py::test_ttl_expiration FAILED                                                                                                          [100%]

=============================================================================== FAILURES ===============================================================================
_________________________________________________________________________ test_ttl_expiration __________________________________________________________________________

dynamodb = dynamodb.ServiceResource()

    @pytest.mark.veryslow
    def test_ttl_expiration(dynamodb):
        duration = 1200 if is_aws(dynamodb) else 3
        # delta is a quarter of the test duration, but no less than one second,
        # and we use it to schedule some expirations a bit after the test starts,
        # not immediately.
        delta = math.ceil(duration / 4)
        assert delta >= 1
        with new_test_table(dynamodb,
            KeySchema=[ { 'AttributeName': 'p', 'KeyType': 'HASH' }, ],
            AttributeDefinitions=[ { 'AttributeName': 'p', 'AttributeType': 'S' } ]) as table:
            # Insert one expiring item *before* enabling the TTL, to verify that
            # items that already exist when TTL is configured also get handled.
            p0 = random_string()
            table.put_item(Item={'p': p0, 'expiration': int(time.time())-60})
            # Enable TTL, using the attribute "expiration":
            client = table.meta.client
            ttl_spec = {'AttributeName': 'expiration', 'Enabled': True}
            response = client.update_time_to_live(TableName=table.name,
                TimeToLiveSpecification=ttl_spec)
            assert response['TimeToLiveSpecification'] == ttl_spec
            # This item should never expire, it is missing the "expiration"
            # attribute:
            p1 = random_string()
            table.put_item(Item={'p': p1})
            # This item should expire ASAP, as its "expiration" has already
            # passed, one minute ago:
            p2 = random_string()
            table.put_item(Item={'p': p2, 'expiration': int(time.time())-60})
            # This item has an expiration more than 5 years in the past (it is my
            # birth date...), so according to the DynamoDB documentation it should
            # be ignored and p3 should never expire:
            p3 = random_string()
            table.put_item(Item={'p': p3, 'expiration': 162777600})
            # This item has as its expiration delta into the future, which is a
            # small part of the test duration, so should expire by the time the
            # test ends:
            p4 = random_string()
            table.put_item(Item={'p': p4, 'expiration': int(time.time())+delta})
            # This item starts with expiration time delta into the future,
            # but before it expires we will move it again, so it will never expire:
            p5 = random_string()
            table.put_item(Item={'p': p5, 'expiration': int(time.time())+delta})
            # This item has an expiration time two durations into the future, so it
            # will not expire by the time the test ends:
            p6 = random_string()
            table.put_item(Item={'p': p6, 'expiration': int(time.time()+duration*2)})
            # Like p4, this item has an expiration time delta into the future,
            # here the expiration time is wrongly encoded as a string, not a
            # number, so the item should never expire:
            p7 = random_string()
            table.put_item(Item={'p': p7, 'expiration': str(int(time.time())+delta)})
            # Like p2, p8 and p9 also have an already-passed expiration time,
            # and should expire ASAP. However, whereas p2 had a straighforward
            # integer like 12345678 as the expiration time, p8 and p9 have
            # slightly more elaborate numbers: p8 has 1234567e1 and p9 has
            # 12345678.1234. Those formats should be fine, and this test verifies
            # the TTL code's number parsing doesn't get confused (in our original
            # implementation, it did).
            p8 = random_string()
            with client_no_transform(table.meta.client) as client:
                client.put_item(TableName=table.name, Item={'p': {'S': p8}, 'expiration': {'N': str((int(time.time())-60)//10) + "e1"}})
            # Similarly, floating point expiration time like 12345678.1 should
            # also be fine (note that Python's time.time() returns floating point).
            # This item should also be expired ASAP too.
            p9 = random_string()
            print(Decimal(str(time.time()-60)))
            table.put_item(Item={'p': p9, 'expiration': Decimal(str(time.time()-60))})
            # We could have just done time.sleep(duration) here, but in case a
            # user is watching this long test, let's output the status every
            # minute, and it also allows us to test what happens when an item
            # p5's expiration time is continuously pushed back into the future:
            start_time = time.time()
            while time.time() < start_time + duration:
                print(f"--- {int(time.time()-start_time)} seconds")
                if 'Item' in table.get_item(Key={'p': p0}):
                    print("p0 alive")
                if 'Item' in table.get_item(Key={'p': p1}):
                    print("p1 alive")
                if 'Item' in table.get_item(Key={'p': p2}):
                    print("p2 alive")
                if 'Item' in table.get_item(Key={'p': p3}):
                    print("p3 alive")
                if 'Item' in table.get_item(Key={'p': p4}):
                    print("p4 alive")
                if 'Item' in table.get_item(Key={'p': p5}):
                    print("p5 alive")
                if 'Item' in table.get_item(Key={'p': p6}):
                    print("p6 alive")
                if 'Item' in table.get_item(Key={'p': p7}):
                    print("p7 alive")
                if 'Item' in table.get_item(Key={'p': p8}):
                    print("p8 alive")
                if 'Item' in table.get_item(Key={'p': p9}):
                    print("p9 alive")
                # Always keep p5's expiration delta into the future
                table.update_item(Key={'p': p5},
                    AttributeUpdates={'expiration': {'Value': int(time.time())+delta, 'Action': 'PUT'}})
                time.sleep(duration/15.0)
    
            # After the delay, p1,p3,p5,p6,p7 should be alive, p0,p2,p4 should not
            assert not 'Item' in table.get_item(Key={'p': p0})
            assert 'Item' in table.get_item(Key={'p': p1})
            assert not 'Item' in table.get_item(Key={'p': p2})
            assert 'Item' in table.get_item(Key={'p': p3})
            assert not 'Item' in table.get_item(Key={'p': p4})
>           assert 'Item' in table.get_item(Key={'p': p5})
E           AssertionError: assert 'Item' in {'ResponseMetadata': {'HTTPHeaders': {'content-length': '2', 'content-type': 'application/x-amz-json-1.0', 'date': 'Wed, 01 Dec 2021 20:21:31 GMT', 'server': 'Seastar httpd'}, 'HTTPStatusCode': 200, 'RetryAttempts': 0}}
E            +  where {'ResponseMetadata': {'HTTPHeaders': {'content-length': '2', 'content-type': 'application/x-amz-json-1.0', 'date': 'Wed, 01 Dec 2021 20:21:31 GMT', 'server': 'Seastar httpd'}, 'HTTPStatusCode': 200, 'RetryAttempts': 0}} = <bound method ResourceFactory._create_action.<locals>.do_action of dynamodb.Table(name='alternator_Test_1638390088828')>(Key={'p': '1OOJ8VVY7J'})
E            +    where <bound method ResourceFactory._create_action.<locals>.do_action of dynamodb.Table(name='alternator_Test_1638390088828')> = dynamodb.Table(name='alternator_Test_1638390088828').get_item

test/alternator/test_ttl.py:310: AssertionError
------------------------------------------------------------------------- Captured stdout call -------------------------------------------------------------------------
fixture creating new table alternator_Test_1638390088828
1638390028.9187481
--- 0 seconds
p0 alive
p1 alive
p2 alive
p3 alive
p4 alive
p5 alive
p6 alive
p7 alive
p8 alive
p9 alive
--- 0 seconds
p1 alive
p3 alive
p6 alive
p7 alive
--- 0 seconds
p1 alive
p3 alive
p5 alive
p6 alive
p7 alive
--- 0 seconds
p1 alive
p3 alive
p5 alive
p6 alive
p7 alive
--- 0 seconds
p1 alive
p3 alive
p5 alive
p6 alive
p7 alive
--- 1 seconds
p1 alive
p3 alive
p6 alive
p7 alive
--- 1 seconds
p1 alive
p3 alive
p5 alive
p6 alive
p7 alive
--- 1 seconds
p1 alive
p3 alive
p5 alive
p6 alive
p7 alive
--- 1 seconds
p1 alive
p3 alive
p5 alive
p6 alive
p7 alive
--- 2 seconds
p1 alive
p3 alive
p5 alive
p6 alive
p7 alive
--- 2 seconds
p1 alive
p3 alive
p6 alive
p7 alive
--- 2 seconds
p1 alive
p3 alive
p5 alive
p6 alive
p7 alive
--- 2 seconds
p1 alive
p3 alive
p5 alive
p6 alive
p7 alive
--- 2 seconds
p1 alive
p3 alive
p5 alive
p6 alive
p7 alive
Deleting table alternator_Test_1638390088828
======================================================================= short test summary info ========================================================================
FAILED test/alternator/test_ttl.py::test_ttl_expiration - AssertionError: assert 'Item' in {'ResponseMetadata': {'HTTPHeaders': {'content-length': '2', 'content-type...
========================================================================== 1 failed in 3.48s ===========================================================================
```

I suspect that 3 seconds for a local test is simply too optimistic a value in my environment and it should be bumped to something higher, but I won't be running any more cases today, so I instead leave this issue here as a reminder. Since this test is marked as `veryslow` anyway, perhaps something like 10 seconds would still be acceptable and produce less false negatives?

/cc @nyh

**Comment by psarna** [comment]
Ah, wait, I should have read the comment in the test....
```
# When running against Scylla configured (for testing purposes) to expire
# items with very short delays, "duration" can be set much lower so this
# test will finish in a much more reasonable time.
```

Never mind!

**Comment by nyh** [comment]
But I don't understand, why did you conclude that this is not a real problem and how did that comment say that? In the past I had such errors when the scanning threas died (but i thought i solved those problems). Or, maybe 3 seconds is really not enough?

**Comment by psarna** [comment]
> But I don't understand, why did you conclude that this is not a real problem and how did that comment say that? In the past I had such errors when the scanning threas died (but i thought i solved those problems). Or, maybe 3 seconds is really not enough?

I closed the issue too soon - I assumed that there's a config value that is set in run.py and not set in my local cluster. It's not the case however, so I'll reopen to keep track of this. I'm almost 100% sure that 3 seconds is sometimes simply not enough in my environment.

**Comment by nyh** [comment]
Eventually there should indeed be such a config parameter, but right now there is a hard-coded 1 second sleep. 1 second is much less than 3 seconds, but perhaps our scan of a gazillion tiny ranges has become so slow that it sometimes takes more than 2 seconds...

**Comment by psarna** [comment]
> Eventually there should indeed be such a config parameter, but right now there is a hard-coded 1 second sleep. 1 second is much less than 3 seconds, but perhaps our scan of a gazillion tiny ranges has become so slow that it sometimes takes more than 2 seconds...

Or, more likely, my Scylla received 1/3th of its cpu time, so 1 second of Scylla's runtime actually lasted around 3 seconds of wall clock time

**Comment by psarna** [comment]
I posted a patch on the mailing list - indeed, test_expiration_ttl always needs ~3 seconds to reach its acceptance conditions, so the 3 second deadline sometimes caused a false negative.

**Comment by nyh** [comment]
This was fixed by commit 26288c1a86a7a14e107e20d682bbe56a7e22b87c, so closing.
